util/latlng2place_mapsapi.py

```diff
@@ -0,0 +1,171 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Library that converts a lat/lng into one or more place DCIDs using Maps API.
+
+Currently supports resolution from lat/lng to admin-area2 (county-equivalent)
+places, admin-area1 (state-equivalent) places and countries.
+
+IMPORTANT NOTE: Please do not hardcode the API key in the calling programs,
+instead pass it as arguments while running them.
+
+Example:
+  ll2p = latlng2place_mapsapi.Resolver(api_key='XYZ')
+  try:
+    dcids = ll2p.resolve(12.0, -34.0)
+  except:
+    logging.error('Failed to resolve 12.0, -34.0')
+  ... more calls to llp.resolve() ...
+"""
+
+import os
+import requests
+import sys
+import threading
+
+# Allows the following module imports to work when running as a script
+sys.path.append(
+    os.path.dirname(os.path.dirname(os.path.dirname(
+        os.path.abspath(__file__)))))
+
+import aa_isocode2dcid
+import alpha2_to_dcid
+
+_LOCK = threading.Lock()
+_PLACE_TYPES = ['administrative_area2', 'administrative_area1', 'country']
+_RECON_ROOT = 'https://api.datacommons.org/v1/recon/resolve/id'
+
+
+def _call_rpc(url):
+    return requests.get(url, timeout=30).json()
+
+
+def _placeid2dcid(place_id):
+    resp = requests.post(_RECON_ROOT,
+                         json={
+                             'in_prop': 'placeId',
+                             'out_prop': 'dcid',
+                             'ids': [place_id]
+                         })
+    resp.raise_for_status()
+    resp_json = resp.json()
+    if 'entities' in resp_json and resp_json['entities']:
+        return resp_json['entities'][0]['outIds'][0]
+    return ''
+
+
+class Resolver(object):
+    """Resolves lat lngs to DCIDs."""
+
+    def __init__(self, api_key, cache_file='', cache_only=False):
+        assert api_key, 'Resolver() needs a valid API key'
+        self._api_key = api_key
+        self._cache_file = cache_file
+        self._cache_only = cache_only
+        self._latlng2place = {}
```

Comment:
nit: please document these class variables, and declare _cf here too for readability.

Reply:
Done